Feat(resiliency): Add litmus pod network latency test

uditgaurav opened the pull request:
Signed-off-by: Udit Gaurav <udit.gaurav@mayadata.io>

## Description

- This PR adds a litmus test to check the resiliency of the CNF application when subject to network latency for a certain chaos duration. 

## Issues:
Refs: 
- [Resilience test: is CNF healthy after pod-network-latency Litmus Chaos test? #520](https://github.com/cncf/cnf-conformance/issues/520)
- EPIC to add various litmus probes: https://github.com/cncf/cnf-conformance/issues/499

## How has this been tested:
 - [ ] Covered by existing integration testing
 - [ ] Added integration testing to cover
 - [ ] Verified all A/C passes
     * [ ] develop
     * [ ] master
     * [ ] tag/other branch
 - [X] Test environment
    * [ ] Shared Packet K8s cluster
    * [ ] New Packet K8s cluster
    * [X] Kind cluster
 - [ ] Have not tested

## Types of changes:
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [X] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - [ ] Documentation update

## Checklist:
**Documentation**
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] No updates required.

**Code Review**
- [X] Does the test handle fatal exceptions, ie. rescue block

**Issue**
- [X] Tasks in issue are checked off

williscool commented:
@uditgaurav this looks awesome thanks so much for the pr!

let me on the team know if you need help getting the test to pass or with rebasing on master.

Aside from that looks good to go

williscool commented:
@uditgaurav @nupejosh also reminded me that we want to get some spec tests for this as well

For inspriation of a spec check out

https://github.com/cncf/cnf-conformance/blob/master/spec/workload/resilience/container_chaos_spec.cr

very simple just installs a sample cnf  and then you run your test

uditgaurav commented:
Sure @williscool will add it. Thanks!

taylor reviewed (commented):
lgtm. thx @uditgaurav

agentpoyo reviewed (commented):
LGTM :+1: and passes individual testing. I updated the `utils_spec.cr` so this would pass running a `crystal spec`

@williscool @wavell @denverwilliams might want to look at the other checks that are failing.

Definitely want to get a cleanup action going, the test fails if running consecutively unless one goes back to delete the pods manually.
Also will need to be added to the workload and resilience options when running.

ksatchit commented on src/tasks/litmuschaos_setup.cr:
```diff
@@ -0,0 +1,82 @@
+
+desc "Install LitmusChaos"
+task "install_litmus" do |_, args|
+    litmus_install = `kubectl apply -f https://raw.githubusercontent.com/litmuschaos/litmus/master/docs/litmus-operator-v1.9.1.yaml`
```
//TODO: Move to 1.11

uditgaurav replied:
Updated with the latest version of litmus.

ksatchit commented on src/tasks/workload/resilience.cr:
```diff
@@ -163,6 +163,46 @@ task "chaos_container_kill", ["install_chaosmesh", "retrieve_manifest"] do |_, a
+    puts "#{destination_cnf_dir}"
+    LOGGING.info "destination_cnf_dir #{destination_cnf_dir}"
+    deployment = Totem.from_file "#{destination_cnf_dir}/manifest.yml"
+    install_experiment = `kubectl apply -f https://hub.litmuschaos.io/api/chaos/1.9.1?file=charts/generic/pod-network-latency/experiment.yaml`
+    install_rbac = `kubectl apply -f https://hub.litmuschaos.io/api/chaos/1.9.1?file=charts/generic/pod-network-latency/rbac.yaml`
```
//TODO: Move to 1.11

uditgaurav replied:
updated with the latest version of litmus.

ksatchit commented:
@agentpoyo thanks for checking. Yes, agree that a clean-up action is necessary and a litmus_cleanup.cr can be added. 

Having said that, the possible reason you are seeing the consecutive tests fail (2nd run onwards) could be on account of crashing coredns pods. In 1.9.x/1.10.x (which the scripts in this PR use), the network-chaos tests for containerd runtime inject the netem rules for egress delay/loss based on the app/target container ID. In cases where liveness probes are configured, a restart of the app containers results in changed IDs making subsequent chaos revert operations unsuccessful. This behaviour has been fixed in the upcoming 1.11 release where we use the PIDs of the corresponding sandbox container (which shares the network namespace with the target app/pod) to inject/revert the netem rules successfully. This will ensure that the coredns pods will recover after the chaos_duration w/o manual intervention (i.e., deleting the pods). We did touch upon this on our previous cnf-dev call, @uditgaurav will be making the necessary version changes in the [litmuschaos_setup.cr](https://github.com/cncf/cnf-conformance/pull/502/files#r542140782) & [resilience.cr](https://github.com/cncf/cnf-conformance/pull/502/files#r542141054) once we make this release post-Dec. 15. 

Coming back to the topic on cleanup-action would like to hear from the team on if the cleanup needs to be executed after each experiment OR if it is something that needs to be done at a suite-level. 

Also, just wanted to bring to your notice that the litmus chaos experiments can be executed with the operator deployed in the "namespaced" mode, where it can execute pod-level chaos experiments against apps in a particular namespace alone (which is where the operator is also installed) v/s the standard mode of execution wherein a single operator oversees and acts on the chaosengine CRs created across namespaces (this is the default & litmuschaos_setup.cr currently does this). AFAIK, the former is seen more in multi-tenant environments where accidental chaos needs to be avoided, while the latter is more common in CI/CD pipelines.

williscool commented:
@uditgaurav great questions

1. On the bug fixes from the 1.11 release: Awesome sounds good looking forward to that hopefully it addresss the issues
2. On the cleanup. The pattern we've generally gone with is doing cleanup after each individual test first. Then refactoring out to suite level if it provides benefits (i.e. faster perf from only doing an install once.)  @wavell and @denverwilliams might chime in with opinions on this also.
3. As far namespaced vs default. I would go with the namespaced but have the default namespace be `default` and allow it to be changed as a command line option
Imagining  we could be testing on multitenant setups and sounds best not to have chaos everywhere (no pun intended lol) but only where explicitly desired. 

... but it also could be true that someone testing a cnf could have multiple components across namespaces they would like choas tested. 

So I would have it as a command lline option maybe something like


``` bash
./cnf-conformance pod-network-latency all_namespaces=true verbose 
```

And have all_namespaces default to false

and for single namespace

``` bash
./cnf-conformance pod-network-latency namespace=another_one verbose 
```

to change from default

thoughts @taylor @wavell @denverwilliams

ksatchit commented:
Thanks @wavell for clarifying.

uditgaurav commented:
Hi @williscool @agentpoyo, The PR is now updated with the latest version of litmus including all required fixes.
- For the multiple iterations of the test please check the logs [here](https://github.com/cncf/cnf-conformance/issues/520#issuecomment-749718410). 
- Added `src/tasks/litmus_cleanup.cr` which includes cleanup task. It is used in `src/tasks/cleanup.cr` and also in spec test (at `spec/workload/resilience/pod_network_latency_spec.cr`).
- As discussed will be adding a namespace scope operator soon.

agentpoyo commented:
> Hi @williscool @agentpoyo, The PR is now updated with the latest version of litmus including all required fixes.
> 
> * For the multiple iterations of the test please check the logs [here](https://github.com/cncf/cnf-conformance/issues/520#issuecomment-749718410).
> * Added `src/tasks/litmus_cleanup.cr` which includes cleanup task. It is used in `src/tasks/cleanup.cr` and also in spec test (at `spec/workload/resilience/pod_network_latency_spec.cr`).
> * As discussed will be adding a namespace scope operator soon.

Very nice. I'll try and do some review/tests tomorrow of the changes. If not then, definitely early next week after the holiday weekend.

wavell commented:
I pulled all of this code into a separate branch, ran it with the additional CI secrets, and it passed: https://github.com/cncf/cnf-conformance/actions/runs/440865672
@agentpoyo @taylor @lixuna
